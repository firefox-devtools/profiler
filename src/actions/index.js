/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
