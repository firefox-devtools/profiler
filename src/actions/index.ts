/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
