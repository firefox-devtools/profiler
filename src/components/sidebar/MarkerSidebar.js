/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Localized } from '@fluent/react';

import explicitConnect from 'firefox-profiler/utils/connect';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import { getSelectedThreadsKey } from 'firefox-profiler/selectors/url-state';
import { TooltipMarker } from 'firefox-profiler/components/tooltip/Marker';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';
import type { ThreadsKey, Marker, MarkerIndex } from 'firefox-profiler/types';

type StateProps = {
  +selectedThreadsKey: ThreadsKey,
  +marker: Marker | null,
  +markerIndex: MarkerIndex | null,
};

type Props = ConnectedProps<{}, StateProps, {}>;

class MarkerSidebarImpl extends React.PureComponent<Props> {
  render() {
    const { marker, markerIndex, selectedThreadsKey } = this.props;

    if (marker === null || markerIndex === null) {
      return (
        <div className="sidebar sidebar-marker-table">
          <Localized id="MarkerSidebar--select-a-marker">
            <div className="sidebar-contents-wrapper">
              Select a marker to display some information about it.
            </div>
          </Localized>
        </div>
      );
    }

    return (
      <aside className="sidebar sidebar-marker-table">
        <div className="sidebar-contents-wrapper">
          <TooltipMarker
            markerIndex={markerIndex}
            marker={marker}
            threadsKey={selectedThreadsKey}
            restrictHeightWidth={false}
          />
        </div>
      </aside>
    );
  }
}

export const MarkerSidebar = explicitConnect<{}, StateProps, {}>({
  mapStateToProps: (state) => ({
    marker: selectedThreadSelectors.getSelectedMarker(state),
    markerIndex: selectedThreadSelectors.getSelectedMarkerIndex(state),
    selectedThreadsKey: getSelectedThreadsKey(state),
  }),
  component: MarkerSidebarImpl,
});
