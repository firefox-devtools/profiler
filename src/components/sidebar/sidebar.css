.sidebar {
  overflow: auto;
  border-top: 1px solid var(--grey-30);
  background: var(--grey-10);
  font-size: 11px;

  --sidebar-indent: 16px;
}

.sidebar-contents-wrapper {
  /* Allow absolute positioning that is dependent of the scrolling. */
  position: relative;
  margin: 16px;
  line-height: 1.5;

  /* Subtract the splitter size. */
  margin-inline-start: calc(16px - 5px);
}

.sidebar-calltree .sidebar-contents-wrapper {
  display: grid;
  align-content: start; /* the grid isn't vertically stretched */
  align-items: center;
  grid-gap: 2px 5px;
  grid-template-columns:
    /* Label */
    1fr
    /* Value */
    min-content
    /* Percentage */
    min-content;
}

.sidebar-titlegroup,
.sidebar-title2,
.sidebar-title3 {
  min-width: 0; /* With this, the cell's minimum size is 0 instead of min-content */
  grid-column-start: span 3;
}

/* Truncate the text with an ellipse */
.sidebar-label,
.sidebar-title2,
.sidebar-title3,
.sidebar-title-label > :nth-child(2) {
  overflow: hidden;
  min-width: 0;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.sidebar-title-label {
  display: grid;
  grid-gap: 5px;
  grid-template-columns: min-content 1fr;
}

.sidebar-title-label > :nth-child(2) {
  color: var(--grey-50);
  text-align: end;
}

.sidebar-titlegroup {
  margin: 0 0 1em;
  line-height: unset;
}

.sidebar-title,
.sidebar-subtitle {
  padding: 0;
  margin: 0;
  font-weight: normal;
}

/* This will be used on both the title and the subtitle */
.can-select-content-input {
  all: unset;
  width: 100%;
  text-overflow: ellipsis;
}

.can-select-content-input:focus-visible {
  text-overflow: unset;
}

.sidebar-title {
  margin-bottom: 0.3em;
  font-size: 1.2em;
  font-weight: 500;
}

.sidebar-subtitle {
  color: var(--grey-50);
  font-size: inherit; /* reset browser styles */
}

.sidebar-title2,
.sidebar-title3 {
  font-weight: normal;
}

.sidebar-title2 {
  margin: 0.9em 0 0.4em;
}

.sidebar-title3 {
  margin: 0.7em 0 0.3em;
}

.sidebar-label {
  color: var(--grey-50);
  grid-column-start: 1;
  margin-inline-start: var(--sidebar-indent);
  white-space: nowrap;
}

.sidebar-label-indent {
  margin-inline-start: calc(var(--sidebar-indent) * 2);
}

.sidebar-value,
.sidebar-percentage {
  text-align: end;
  white-space: nowrap;
}

.sidebar-histogram-bar {
  background-color: var(--grey-30);
  grid-column-start: span 3;
  margin-inline-start: var(--sidebar-indent);
}

.sidebar-histogram-bar-color {
  height: 5px;
}

.sidebar-histogram-bar-color.category-color-transparent {
  /* A white bar does not show up well here */
  background-color: var(--grey-50);
}

.sidebar-histogram-bar-color.category-color-grey {
  /* The grey category color is the same our background color, choose a darker grey instead. */
  background-color: var(--grey-50);
}

.sidebar-toggle {
  padding: 0;
  border: 0;
  background: 0;
  color: var(--grey-50);
  font-size: inherit;
}

/* Opt out of the default photon styling, as it doesn't look very nice with the
   overflow: hidden of the container */
.sidebar-toggle:focus-visible {
  border: 0;
  border-radius: 2px;
  box-shadow: none;
  color: var(--blue-60);
  text-decoration: underline;
}

/* This is the twisty arrow */
.sidebar-toggle::before {
  position: absolute;
  width: 0;
  height: 0;
  border-style: solid;
  border-color: transparent;
  margin-top: 1px;
  background: none;
  border-block-width: 5px;
  border-inline-end-width: 4px;
  border-inline-start-color: currentcolor;
  border-inline-start-width: 8px;
  content: '';

  /* Make sure the arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
  margin-inline-start: -11px;
}

.sidebar-toggle:active:hover {
  color: var(--grey-60);
}

.sidebar-toggle.expanded::before {
  border-width: 8px 5px 4px;
  border-style: solid;
  border-color: currentcolor transparent transparent;
  margin-top: 3px;
  margin-inline-start: -12px;
}

@media (forced-colors: active) {
  .sidebar-toggle,
  .sidebar-toggle::before {
    color: ButtonText;
  }

  .sidebar-label:hover .sidebar-toggle,
  .sidebar-label:hover .sidebar-toggle::before {
    color: SelectedItem;
  }
}
