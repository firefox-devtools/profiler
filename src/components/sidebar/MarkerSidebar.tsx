/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import * as React from 'react';
import { Localized } from '@fluent/react';

import explicitConnect from 'firefox-profiler/utils/connect';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getSelectedThreadsKey,
  getSelectedTab,
} from 'firefox-profiler/selectors/url-state';
import { TooltipMarker } from 'firefox-profiler/components/tooltip/Marker';
import { updateBottomBoxContentsAndMaybeOpen } from 'firefox-profiler/actions/profile-view';
import { getBottomBoxInfoForStackFrame } from 'firefox-profiler/profile-logic/profile-data';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';
import type {
  ThreadsKey,
  Marker,
  MarkerIndex,
  IndexIntoStackTable,
  Thread,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';

type StateProps = {
  readonly selectedThreadsKey: ThreadsKey;
  readonly marker: Marker | null;
  readonly markerIndex: MarkerIndex | null;
  readonly thread: Thread;
  readonly selectedTab: TabSlug;
};

type DispatchProps = {
  readonly updateBottomBoxContentsAndMaybeOpen: typeof updateBottomBoxContentsAndMaybeOpen;
};

type Props = ConnectedProps<{}, StateProps, DispatchProps>;

class MarkerSidebarImpl extends React.PureComponent<Props> {
  _onStackFrameClick = (stackIndex: IndexIntoStackTable) => {
    const { thread, selectedTab, updateBottomBoxContentsAndMaybeOpen } =
      this.props;
    const bottomBoxInfo = getBottomBoxInfoForStackFrame(stackIndex, thread);
    updateBottomBoxContentsAndMaybeOpen(selectedTab, bottomBoxInfo);
  };

  override render() {
    const { marker, markerIndex, selectedThreadsKey } = this.props;

    if (marker === null || markerIndex === null) {
      return (
        <div className="sidebar sidebar-marker-table">
          <Localized id="MarkerSidebar--select-a-marker">
            <div className="sidebar-contents-wrapper">
              Select a marker to display some information about it.
            </div>
          </Localized>
        </div>
      );
    }

    return (
      <aside className="sidebar sidebar-marker-table">
        <div className="sidebar-contents-wrapper">
          <TooltipMarker
            markerIndex={markerIndex}
            marker={marker}
            threadsKey={selectedThreadsKey}
            restrictHeightWidth={false}
            onStackFrameClick={this._onStackFrameClick}
          />
        </div>
      </aside>
    );
  }
}

export const MarkerSidebar = explicitConnect<{}, StateProps, DispatchProps>({
  mapStateToProps: (state) => ({
    marker: selectedThreadSelectors.getSelectedMarker(state),
    markerIndex: selectedThreadSelectors.getSelectedMarkerIndex(state),
    selectedThreadsKey: getSelectedThreadsKey(state),
    thread: selectedThreadSelectors.getThread(state),
    selectedTab: getSelectedTab(state),
  }),
  mapDispatchToProps: {
    updateBottomBoxContentsAndMaybeOpen,
  },
  component: MarkerSidebarImpl,
});
