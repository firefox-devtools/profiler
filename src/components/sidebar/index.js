/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';

import { CallTreeSidebar } from './CallTreeSidebar';
import { MarkerSidebar } from './MarkerSidebar';

import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';

import './sidebar.css';

export function selectSidebar(
  selectedTab: TabSlug
): React.ComponentType<{||}> | null {
  return {
    calltree: CallTreeSidebar,
    'function-list': CallTreeSidebar,
    'flame-graph': CallTreeSidebar,
    'stack-chart': null,
    'marker-chart': null,
    'marker-table': MarkerSidebar, // MarkerSidebar
    'network-chart': null,
    'js-tracer': null,
  }[selectedTab];
}
