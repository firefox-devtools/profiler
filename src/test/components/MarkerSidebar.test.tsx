/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import { Provider } from 'react-redux';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import { MarkerSidebar } from '../../components/sidebar/MarkerSidebar';
import { changeSelectedMarker } from '../../actions/profile-view';

import { storeWithProfile } from '../fixtures/stores';
import { getMarkerTableProfile } from '../fixtures/profiles/processed-profile';

describe('MarkerSidebar', function () {
  function setup() {
    const profile = getMarkerTableProfile();
    const store = storeWithProfile(profile);

    const renderResult = render(
      <Provider store={store}>
        <MarkerSidebar />
      </Provider>
    );
    return {
      ...renderResult,
      ...store,
      profile,
    };
  }

  it('matches the snapshots when displaying data about the currently selected node', () => {
    const { dispatch, profile, container } = setup();

    act(() => {
      dispatch(
        changeSelectedMarker(
          0,
          profile.threads[0].markers.data.findIndex(
            (data) => data && data.type === 'IPC'
          )
        )
      );
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
