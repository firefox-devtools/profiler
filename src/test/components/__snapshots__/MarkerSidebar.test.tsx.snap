// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MarkerSidebar matches the snapshots when displaying data about the currently selected node 1`] = `
<aside
  class="sidebar sidebar-marker-table"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <div
      class="tooltipMarker"
    >
      <div
        class="tooltipHeader"
      >
        <div
          class="tooltipOneLine"
        >
          <div
            class="tooltipTiming"
          >
            unknown duration
          </div>
          <div
            class="tooltipTitle"
          >
            <span
              class="tooltipTitleText"
            >
              IPC — sent to 2222
            </span>
            <button
              aria-label="Only show markers matching: “⁨IPCOut⁩”"
              class="tooltipTitleFilterButton"
              title="Only show markers matching: “⁨IPCOut⁩”"
              type="button"
            />
          </div>
        </div>
      </div>
      <div
        class="tooltipDetails"
      >
        <div
          class="tooltipLabel"
        >
          Type
          :
        </div>
        PContent::Msg_PreferenceUpdate
        <div
          class="tooltipLabel"
        >
          Sync
          :
        </div>
        false
        <div
          class="tooltipLabel"
        >
          From
          :
        </div>
        Empty (Thread ID: 0)
        <div
          class="tooltipLabel"
        >
          Other Pid
          :
        </div>
        PID: 2222
        <div
          class="tooltipLabel"
        >
          Send Thread Latency
          :
        </div>
        unknown
        <div
          class="tooltipLabel"
        >
          IPC Send Duration
          :
        </div>
        unknown
        <div
          class="tooltipLabel"
        >
          IPC Recv Latency
          :
        </div>
        unknown
        <div
          class="tooltipLabel"
        >
          Recv Thread Latency
          :
        </div>
        unknown
        <div
          class="tooltipLabel"
        >
          Track
          :
        </div>
        Empty
      </div>
    </div>
  </div>
</aside>
`;
